### PR TITLE
added state and multilevel prolong + tests.

### PR DESCRIPTION
**Pull request by alsgregory**
Initial code commit. Including:
- travis, make and setup
- **init**.py and the source code for the state object
- methods for recursive (multiple level) prolonging and injecting of functions in a hierarchy
- tests for these
- demos

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+install:
+  - curl -O https://raw.githubusercontent.com/firedrakeproject/firedrake/master/scripts/firedrake-install
+  - python ./firedrake-install --disable-ssh --minimal-petsc
+  - . ./firedrake/bin/activate
+  - python setup.py install
+
+script:
+  - make lint
+  - make test
+
```
Comment (miklos1) [inline .travis.yml]: please...

Comment (alsgregory) [inline .travis.yml]: Noted. Isn't it ironic that the script checking linting errors isn't checked for this itself...

Comment (alsgregory) [inline .travis.yml]: done on new commit

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
+notifications:
+  slack:
+    secure: cuB9waMrZiRnJjJLq68NTSg1nB+ypbCTy4cphu0VjfuDYhmQc4MgnkvmzLntsE9X9Oj8kSP51rRjD+IJJ8qT1GTMi/5XAqKNzncONEdx5T0Tict0xxAaY8OcumOV68tisQ1jEeAVA65zcvPoyWOVv+ha4MdHLx8+J85h/zOFMRALf2tJ8tUnC2iHidWCYsQS8vQEkKp1eDWJRdcJY+OrmeV5tlgWXP56b6pRdnhoWwOkwtRols6yoPw1j6uNewb3N2uvJIIw5g7ji209nXFM+/HFGLLqNlWUpqickbaZEG/m5VqeuxeaZTl6jlW2w2+B5zyMPKR8NHeSQ8+c6Si08LbgQDGkLh1xHNQOTxqTEOHUOsLgBm7Y+NX1QDfIfclvY+dhWhUiNHVT39+om/I/8MvpuDLLtujijJIq+mluJdjuAkWWXCVqYGMll0SalvEQf43Os6yxZnewJ0QjlajUY8txgPR9yyVLQduddiMymSIwAAKE3gLlhTpl59W3kRcKcd0tTUtnqH7KAg6GBnFqkqnUPv9Fmws+eOvyThg36gX7+RWbo/mwQZCH265qY6cf02pbsU9Hoo5HTXd7nWi4G/NL3vKR3yySrOCzfzgwY2vozwrWY09pKphNtRhtLZkXEKdvDyFt9i6/nh6muiaZEQ22dxxK/1iBKTeR+IW5p48=
```
Comment (wence-) [inline .travis.yml]: This is the wrong secure string.  Remove the notifications section and I will add it in a separate pull request.

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+before_install:
+  - pip install -U pip
+  - pip install -U virtualenv
+  - pip install flake8
+
+install:
+  - curl -O https://raw.githubusercontent.com/firedrakeproject/firedrake/master/scripts/firedrake-install
+  - python ./firedrake-install --disable-ssh --minimal-petsc
+  - . ./firedrake/bin/activate
+  - python setup.py install
```
Comment (wence-) [inline .travis.yml]: I would run `pip install .` instead.

**Comment by alsgregory** [comment]
I've changed travis file for those corrections. I've also just changed the state.py slightly to remove a setting where one could carry lists of Functions. This over complicated things for now. Is it ok to merge after this?

**Comment by alsgregory** [comment]
Major update to initial commit. After conversations about prolonging and injecting over recursive levels (this functionality has been pushed to a branch in firedrakeproject/firedrake). 

Removed:
- prolonging and injecting over multiple levels is removed and incorporated into multigrid

Added:
- Generalised hierarchies are able to build FS and Mesh hierarchies that are refined by a factor greater than two.
- tests and demos

Altered:
- no more dependence on FunctionHierarchy, which will be removed from multigrid
- state.py removing unnecessary error checks

**Comment by alsgregory** [comment]
Can I merge?

**Pull request merged**